### PR TITLE
[DO NOT MERGE] Exec into the container to get bind creds

### PR DESCRIPTION
**Pull request by rthallisey**
Instead of getting bind credentials from stdout, the broker will exec into the container and get them.

More details in this patch: https://github.com/fusor/ansible-playbook-bundle-library/pull/1

### CODE REVIEW

File: pkg/apb/bind.go [hunk 1]
```diff
@@ -86,8 +83,24 @@ func getPodName(output []byte, log *logging.Logger) (string, error) {
...
-	return runCommand("oc", "logs", "-f", podname)
+func monitorOutput(podname string, mon chan []byte, log *logging.Logger) {
+	// TODO: Error handling here
+	// It would also be nice to gather the script output that exec runs
+	// instead of only getting the credentials
+	retries := 5
+	for r := 1; r <= retries; r++ {
+		output, _ := runCommand("oc", "exec", podname, "broker-bind-creds")
+
+		if strings.Contains(string(output), "BIND_CREDENTIALS") {
```
Comment (jmrodri) [inline pkg/apb/bind.go]: Didn't `Contains` existed. Not sure how I missed that before. +1

File: pkg/apb/bind.go [hunk 2]
```diff
@@ -86,8 +83,24 @@ func getPodName(output []byte, log *logging.Logger) (string, error) {
...
+			mon <- output
+			close(mon)
+			return
+		}
+
+		time.Sleep(time.Duration(r) * time.Second)
+		log.Warning("Retry attempt %d: exec into %s failed", r, podname)
```
Comment (jmrodri) [inline pkg/apb/bind.go]: Good log.

File: pkg/apb/bind.go [hunk 3]
```diff
@@ -51,22 +51,19 @@ func Bind(
...
-		return buildBindData(bindout)
+	out := make(chan []byte)
+	go func() {
+		monitorOutput(podname, out, log)
+	}()
+
+	var bindout []byte
+	for msg := range out {
+		log.Info(string(msg))
+		bindout = msg
```
Comment (jmrodri) [inline pkg/apb/bind.go]: So we monitor the output in a go routine, then we wait for there to be something in the channel to parse. I like it.

Comment (jmrodri) [inline pkg/apb/bind.go]: If the monitor fails to write anything to the channel, what happens? Does this hang? does it quit?

Comment (rthallisey) [inline pkg/apb/bind.go]: If the monitor writes nothing to the channel we'll hang.

File: pkg/apb/bind.go [hunk 4]
```diff
@@ -86,8 +83,24 @@ func getPodName(output []byte, log *logging.Logger) (string, error) {
...
+		if strings.Contains(string(output), "BIND_CREDENTIALS") {
+			mon <- output
+			close(mon)
+			return
+		}
+
+		time.Sleep(time.Duration(r) * time.Second)
+		log.Warning("Retry attempt %d: exec into %s failed", r, podname)
+	}
+	return
```
Comment (rthallisey) [inline pkg/apb/bind.go]: Need to send a timeout error message to the channel and ```close(mon)```. Return here isn't enough.

**Comment by jmrodri** [comment]
@rthallisey we're going to hold off on this PR because of the work @eriknelson has been doing recently things have changed. I want to keep this open because we want some of this work but it will not be a clean merge.

**Comment by rthallisey** [comment]
No rush.  I'll rebase when we're ready to merge.

**Comment by jmrodri** [comment]
Now I don't know how this will be affected with all the new `ext_creds.go` refactor we did earlier.

**Comment by jmrodri** [comment]
@rthallisey rebase this and we can take another look at it. We also are getting bind credentials from the output of a **provision** call now.

**Comment by eriknelson** [comment]
Because credentials can now come out of a provision or a bind call, we need to chat about the various scenarios that could arise and how to react to each. Ex:

```
Credentials returned:
prov: T, bind: F -> pass back prov creds from bind, we do this today.
prov: F, bind: T -> pass back bind creds, basically the same as the above.
prov: T, bind: T -> First, is this okay? If so, how do we handle this? Merge one onto the other? Which 
  takes precendence if there is a key conflict?
prov: F, bind F -> Is this okay? I lean no. If a bind is basically defined by the return of credentials, if 
  there are none available, this is bad. Can we think of a case where a bind makes sense that 
  *doesn't* return anything to the consuming application? NOTE: This currently segfaults. Needs to get 
   fixed and tested against our decision here.
```

**Comment by rthallisey** [comment]
Ya let's revisit this method.

**Comment by eriknelson** [comment]
Related #104

**Comment by rthallisey** [comment]
https://github.com/fusor/ansible-service-broker/pull/124

**Pull request closed without merging**